spec/controllers/frontend_log_controller_spec.rb

```diff
@@ -0,0 +1,80 @@
+require 'rails_helper'
+
+describe FrontendLogController do
+  describe '#create' do
+    subject(:action) { post :create, params: params }
+
+    let(:user) { create(:user, :with_phone, with: { phone: '+1 (202) 555-1212' }) }
+    let(:params) do
+      {
+        event: Analytics::FRONTEND_DOC_AUTH_ASYNC_UPLOAD,
+        payload: { message: 'To be logged...' },
+      }
+    end
+    let(:json) { JSON.parse(response.body, symbolize_names: true) }
+
+    context 'user is signed in' do
+      before do
+        sign_in user
+        stub_analytics
+      end
+
+      it 'succeeds' do
+        expect(@analytics).to receive(:track_event).
+          with(params[:event], params[:payload])
+
+        action
+
+        expect(response).to have_http_status(:ok)
+        expect(json[:success]).to eq(true)
+      end
+
+      context 'unallowed event type' do
+        it 'rejects a request with an event that is not allowed' do
+          expect(@analytics).not_to receive(:track_event)
+
+          params[:event] = 'custom event'
+          action
+
+          expect(response).to have_http_status(:bad_request)
+          expect(json[:success]).to eq(false)
+        end
+      end
+
+      context 'missing a parameter' do
```

Comment:
I was sort of expecting this to fail. It doesn't (I think because `params.permit`?), but probably good to have a test case for our new hash validation:

```suggestion
      context 'invalid param' do
        it 'rejects a non-hash payload' do
          expect(@analytics).not_to receive(:track_event)

          params[:payload] = 'abc'
          action

          expect(response).to have_http_status(:bad_request)
          expect(json[:success]).to eq(false)
        end
      end

      context 'missing a parameter' do
```

Reply:
I included this and added a spec for a non-string event.